packages/components/src/image-cropper/reducer.ts

```diff
@@ -0,0 +1,423 @@
+/**
+ * Internal dependencies
+ */
+import type { ResizeDirection, Position } from './types';
+import {
+	rotatePoint,
+	degreeToRadian,
+	calculateRotatedBounds,
+	getMinScale,
+} from './math';
+
+export type State = {
+	// The image dimensions.
+	image: {
+		// The x position of the image center.
+		x: number;
+		// The y position of the image center.
+		y: number;
+		// The width of the image.
+		width: number;
+		// The height of the image.
+		height: number;
+	};
+	// The image transforms.
+	transforms: {
+		// The angle of the image in degrees, from -45 to 45 degrees.
+		angle: number;
+		// The number of 90-degree rotations clockwise.
+		rotations: number;
+		// The image scale.
+		scale: { x: number; y: number };
+	};
+	// The cropper window dimensions.
+	cropper: {
+		// The x position of the cropper window center.
+		x: number;
+		// The y position of the cropper window center.
+		y: number;
+		// The width of the cropper window.
+		width: number;
+		// The height of the cropper window
+		height: number;
+	};
+	// Whether the cropper window is resizing.
+	isResizing: boolean;
+	// Whether the image is dragging/moving.
+	isDragging: boolean;
+	// Whether the image is zooming/pinching.
+	isZooming: boolean;
+};
+
+type Action =
+	// Zoom in/out to a scale.
+	| { type: 'ZOOM'; scale: number; position: Position }
+	// End zooming.
+	| { type: 'ZOOM_END' }
+	// Flip the image horizontally.
+	| { type: 'FLIP' }
+	// Rotate the image to an angle.
+	| { type: 'ROTATE'; angle: number }
+	// Rotate the image 90-degree clockwise or counter-clockwise.
+	| { type: 'ROTATE_CLOCKWISE'; isCounterClockwise?: boolean }
+	// Move the image to a position.
+	| { type: 'MOVE'; x: number; y: number }
+	// End moving the image.
+	| { type: 'MOVE_END' }
+	// Start resizing the cropper window.
+	| { type: 'RESIZE_START' }
+	// Resize the cropper window by a delta size in a direction.
+	| { type: 'MOVE_WINDOW'; x: number; y: number }
+	// Resize the cropper window by a delta size in a direction.
+	| {
+			type: 'RESIZE_WINDOW';
+			direction: ResizeDirection;
+			delta: { width: number; height: number };
+	  }
+	// Resize the container and image to a new width.
+	| { type: 'RESIZE_CONTAINER'; width: number }
+	// Reset the state to the initial state.
+	| { type: 'RESET' };
+
+function createInitialState( {
+	width,
+	height,
+}: {
+	width: number;
+	height: number;
+} ): State {
+	return {
+		image: {
+			width,
+			height,
+			x: 0,
+			y: 0,
+		},
+		transforms: {
+			angle: 0,
+			rotations: 0,
```

Comment:
Out of interest, why does the angle needs to be stored separately to rotations? The two feel like the same thing to me.

Reply:
They are the same thing, but this matches the UI of the Google photos cropper that many of these interactions seemed to be based off of. One could deconstruct any angle into these two parts, but I haven't convinced myself which way would be better. Mostly thinking in terms of redo/undo.

<img width="294" alt="image" src="https://github.com/user-attachments/assets/716b366b-5f9c-434d-b975-9902d238526d">

Reply:
Yep! We need two separate values here so that we can distinguish between `angle: 45, rotations: 0` and `angle: -45, rotations: 1`, which are both 45 degree but represent different states in the UI.